index.js

```diff
@@ -1,301 +1,319 @@
+/**
+ * @private
+ */
+function _rotate(originalPath, suffix, stream) {
 
-            // remove the current path from the queue
-            that._compressq.shift();
+    let newPath = originalPath + '.' + suffix
+    stream.emit('rotate', originalPath, newPath)
 
-            // if it's not empty, continue to the next one
-            if ( that._compressq.length ) {
-                that._compress();
-            }
-
-            // remove the uncompressed file
-            that._unlink( path )
-        });
+    return fs.renameAsync(originalPath, newPath)
+        .then(() => {
+            return _compress(newPath, stream)
+        })
+        .then((gZipedPath) => {
+            return _upload(gZipedPath, stream)
+        })
+        .tap((gZipedPath) => {
+            stream.emit('rotated', originalPath, gZipedPath || '')
+        })
 }
 
-RotateStream.prototype._upload = function ( path ) {
-    var that = this;
-    if ( path ) {
-        this._uploadq.push( path );
-        if ( this._uploadq.length > 1 ) {
-            return; // upload already running
-        }
-    }
+/**
+ * @private
+ */
+function _compress(path, stream) {
+    return new Promise((resolve, reject) => {
+        let pathGzip = path + '.gz'
 
-    path = this._uploadq[ 0 ];
-    var date = path.match( /(\d{4})-(\d{2})-(\d{2})T\d{2}:\d{2}:\d{2}\.\d{3}Z/ )
-    var fname = pathlib.basename( path );
-    var s3 = this._s3;
-    var key = pathlib.join( s3.prefix, date[ 1 ], date[ 2 ], date[ 3 ], fname );
-
-    this.emit( 'upload', path, s3.bucket + '/' + key );
-    s3.putObject({
-        Bucket: s3.bucket,
-        Key: key,
-        Body: fs.createReadStream( path )
-    })
-    .on( 'error', this.emit.bind( this, 'error' ) )
-    .on( 'httpUploadProgress', function ( progress ) {
-        that.emit( 'uploading', path, s3.bucket + '/' + key, progress );
-    })
-    .on( 'success', function () {
-        that.emit( 'uploaded', path, s3.bucket + '/' + key );
-
-        // remove the current path from the queue
-        that._uploadq.shift();
-
-        // if it's not empty, continue to the next one
-        if ( that._uploadq.length ) {
-            that._upload();
-        }
-
-        // remove the uploaded file
-        that._unlink( path )
-    })
-    .send()
-}
-
-RotateStream.prototype._unlink = function ( path, cb ) {
-    var that = this;
-    this.emit( 'delete', path );
-    fs.unlink( path, function ( err ) {
-        if ( err ) {
-            that.emit( 'error', err );
+        if (!stream.compress) {
+            resolve(path)
         } else {
-            that.emit( 'deleted', path );
-            if ( cb ) {
-                cb();
-            }
+            stream.emit('compress', path, pathGzip)
+            fs.createReadStream(path)
+                .pipe(zlib.createGzip())
+                .pipe(fs.createWriteStream(pathGzip))
+                .once('error', (err) => {
+                    stream.emit('error', err)
```

Comment:
An error is already being emitted on the main `_write` function. So now it will happen twice for the same error. Choose which part of the code should emit the error, instead of duplicating the error event.

Reply:
Removed duplicate `emit('error', err)` calles